.github/workflows/e2e-macos.yaml

```diff
@@ -82,5 +87,78 @@ jobs:
         run: |
           git status
           git clean -f -d
-          REGISTRY=${{ steps.vars.outputs.has_creds == true && env.REGISTRY || '' }} make ${{ inputs.test-command }}
+          if [[ "${{ github.event_name }}" == "push" ]] && [[ "${{ github.ref }}" == "refs/heads/main" ]] || [[ "${{ github.event_name }}" == "pull_request" ]] && [[ "${{ github.base_ref }}" == "main" ]]; then
+              REGISTRY=${{ steps.vars.outputs.has_creds == true && env.REGISTRY || '' }} make "${{ inputs.test-command }}-with-report"
+          else
+              REGISTRY=${{ steps.vars.outputs.has_creds == true && env.REGISTRY || '' }} make ${{ inputs.test-command }}
+          fi
         shell: zsh {0}
+      - name: Set artifacts name outputs
+        if: always() && (github.event_name == 'push' && github.ref == 'refs/heads/main') || (github.event_name == 'pull_request' && github.base_ref == 'main')
+        id: set-multiple-vars
+        run: |
+          echo "VM_REPORT=${{ github.run_id }}-${{ github.run_attempt }}-e2e-vm-report.json" >> $GITHUB_OUTPUT
+          echo "CONTAINER_REPORT=${{ github.run_id }}-${{ github.run_attempt }}-e2e-container-report.json" >> $GITHUB_OUTPUT
+          echo "VM_SERIAL_REPORT=${{ github.run_id }}-${{ github.run_attempt }}-e2e-vm-serial-report.json" >> $GITHUB_OUTPUT
+      - name: Upload reports artifact
+        if: always() && (github.event_name == 'push' && github.ref == 'refs/heads/main') || (github.event_name == 'pull_request' && github.base_ref == 'main')
+        uses: actions/upload-artifact@v4
```

Comment:
Can you pin this using the action's sha? i.e. actions/upload-artifact@<hash> # v4.x.x